File: acp-sdk-typescript/src/messages.ts
import { randomUUID } from "node:crypto";
import { ACP_VERSION, DEFAULT_CRYPTO_SUITE } from "./constants.js";
import { JsonMap, JsonValue, toJsonMap } from "./jsonSupport.js";
import { validationError } from "./errors.js";

export type MessageClass = "SEND" | "ACK" | "FAIL" | "CAPABILITIES" | "COMPENSATE";
export type DeliveryState =
  | "PENDING"
  | "DELIVERED"
  | "ACKNOWLEDGED"
  | "FAILED"
  | "DECLINED"
  | "EXPIRED";
export type DeliveryMode = "auto" | "direct" | "relay" | "amqp" | "mqtt";

export interface WrappedContentKey {
  recipient: string;
  ephemeral_public_key: string;
  nonce: string;
  ciphertext: string;
}

export interface Envelope {
  acp_version: string;
  message_class: MessageClass;
  message_id: string;
  operation_id: string;
  timestamp: string;
  expires_at: string;
  sender: string;
  recipients: string[];
  context_id: string;
  crypto_suite: string;
  correlation_id?: string;
  in_reply_to?: string;
}

export interface ProtectedPayload {
  nonce: string;
  ciphertext: string;
  wrapped_content_keys: WrappedContentKey[];
  payload_hash: string;
  signature_kid: string;
  signature: string;
}

export interface AcpMessage {
  envelope: Envelope;
  protected: ProtectedPayload;
  sender_identity_document?: JsonMap;
}

export interface DeliveryOutcome {
  recipient: string;
  state: DeliveryState;
  status_code?: number;
  response_class?: MessageClass;
  reason_code?: string;
  detail?: string;
  response_message?: JsonMap;
}

export interface SendResult {
  operation_id: string;
  message_id: string;
  message_ids: string[];
  outcomes: DeliveryOutcome[];
}

export interface CompensateInstruction {
  operation_id: string;
  reason: string;
  actions: JsonMap[];
}

export function parseMessageClass(value: string | undefined): MessageClass | undefined {
  if (!value) {
    return undefined;
  }
  if (
    value === "SEND" ||
    value === "ACK" ||
    value === "FAIL" ||
    value === "CAPABILITIES" ||
    value === "COMPENSATE"
  ) {
    return value;
  }
  return undefined;
}

export function buildEnvelope(input: {
  sender: string;
  recipients: string[];
  message_class: MessageClass;
  context_id: string;
  expires_in_seconds: number;
  operation_id?: string;
  correlation_id?: string;
  in_reply_to?: string;
  crypto_suite?: string;
}): Envelope {
  const now = new Date();
  const expires = new Date(now.getTime() + Math.max(1, input.expires_in_seconds) * 1000);
  const envelope: Envelope = {
    acp_version: ACP_VERSION,
    message_class: input.message_class,
    message_id: randomUUID(),
    operation_id: input.operation_id ?? randomUUID(),
    timestamp: now.toISOString(),
    expires_at: expires.toISOString(),
    sender: input.sender,
    recipients: input.recipients,
    context_id: input.context_id,
    crypto_suite: input.crypto_suite ?? DEFAULT_CRYPTO_SUITE
  };
  if (input.correlation_id) {
    envelope.correlation_id = input.correlation_id;
  }
  if (input.in_reply_to) {
    envelope.in_reply_to = input.in_reply_to;
  }
  validateEnvelope(envelope);
  return envelope;
}

export function validateEnvelope(envelope: Envelope): void {
  if (!envelope.sender.trim()) {
    throw validationError("Envelope sender is required");
  }
  if (!Array.isArray(envelope.recipients) || envelope.recipients.length === 0) {
    throw validationError("Envelope recipients must not be empty");
  }
  const timestamp = Date.parse(envelope.timestamp);
  const expires = Date.parse(envelope.expires_at);
  if (Number.isNaN(timestamp) || Number.isNaN(expires)) {
    throw validationError("Envelope timestamps must be RFC3339 strings");
  }
  if (expires <= timestamp) {
    throw validationError("Envelope expires_at must be after timestamp");
  }
}

export function isExpired(envelope: Envelope): boolean {
  const expires = Date.parse(envelope.expires_at);
  return Number.isNaN(expires) || expires <= Date.now();
}

export function parseAcpMessage(map: JsonMap): AcpMessage {
  const envelope = toJsonMap(map.envelope) as unknown as Envelope;
  const protectedPayload = toJsonMap(map.protected) as unknown as ProtectedPayload;
  const message: AcpMessage = {
    envelope,
    protected: protectedPayload
  };
  if (map.sender_identity_document !== undefined && map.sender_identity_document !== null) {
    message.sender_identity_document = toJsonMap(map.sender_identity_document);
  }
  validateEnvelope(envelope);
  return message;
}

export function messageToMap(message: AcpMessage): JsonMap {
  const output: JsonMap = {
    envelope: message.envelope as unknown as JsonValue,
    protected: message.protected as unknown as JsonValue
  };
  if (message.sender_identity_document) {
    output.sender_identity_document = message.sender_identity_document;
  }
  return output;
}

export function buildAckPayload(receivedMessageId: string, status: string): JsonMap {
  return {
    status,
    received_message_id: receivedMessageId
  };
}

export function buildFailPayload(reasonCode: string, detail: string, retriable: boolean): JsonMap {
  return {
    reason_code: reasonCode,
    detail,
    retriable
  };
}

File: sdks/go/messages.go
/*
 * Copyright 2026 ACP Project
 * Licensed under the Apache License, Version 2.0
 * See LICENSE file for details.
 */

package acp

import (
	"encoding/json"
	"strings"
	"time"

	"github.com/google/uuid"
)

type MessageClass string

const (
	MessageSend         MessageClass = "SEND"
	MessageAck          MessageClass = "ACK"
	MessageFail         MessageClass = "FAIL"
	MessageCapabilities MessageClass = "CAPABILITIES"
	MessageCompensate   MessageClass = "COMPENSATE"
)

type DeliveryState string

const (
	StatePending      DeliveryState = "PENDING"
	StateDelivered    DeliveryState = "DELIVERED"
	StateAcknowledged DeliveryState = "ACKNOWLEDGED"
	StateFailed       DeliveryState = "FAILED"
	StateDeclined     DeliveryState = "DECLINED"
	StateExpired      DeliveryState = "EXPIRED"
)

type DeliveryMode string

const (
	DeliveryAuto   DeliveryMode = "auto"
	DeliveryDirect DeliveryMode = "direct"
	DeliveryRelay  DeliveryMode = "relay"
	DeliveryAMQP   DeliveryMode = "amqp"
	DeliveryMQTT   DeliveryMode = "mqtt"
)

type WrappedContentKey struct {
	Recipient          string `json:"recipient"`
	EphemeralPublicKey string `json:"ephemeral_public_key"`
	Nonce              string `json:"nonce"`
	Ciphertext         string `json:"ciphertext"`
}

type Envelope struct {
	ACPVersion    string       `json:"acp_version"`
	MessageClass  MessageClass `json:"message_class"`
	MessageID     string       `json:"message_id"`
	OperationID   string       `json:"operation_id"`
	Timestamp     string       `json:"timestamp"`
	ExpiresAt     string       `json:"expires_at"`
	Sender        string       `json:"sender"`
	Recipients    []string     `json:"recipients"`
	ContextID     string       `json:"context_id"`
	CryptoSuite   string       `json:"crypto_suite"`
	Namespace     *string      `json:"namespace,omitempty"`
	CorrelationID *string      `json:"correlation_id,omitempty"`
	InReplyTo     *string      `json:"in_reply_to,omitempty"`
}

type ProtectedPayload struct {
	Nonce              string              `json:"nonce"`
	Ciphertext         string              `json:"ciphertext"`
	WrappedContentKeys []WrappedContentKey `json:"wrapped_content_keys"`
	PayloadHash        string              `json:"payload_hash"`
	SignatureKID       string              `json:"signature_kid"`
	Signature          string              `json:"signature"`
}

type AcpMessage struct {
	Envelope               Envelope         `json:"envelope"`
	Protected              ProtectedPayload `json:"protected"`
	SenderIdentityDocument map[string]any   `json:"sender_identity_document,omitempty"`
}

type DeliveryOutcome struct {
	Recipient       string         `json:"recipient"`
	State           DeliveryState  `json:"state"`
	StatusCode      *int           `json:"status_code,omitempty"`
	ResponseClass   *MessageClass  `json:"response_class,omitempty"`
	ReasonCode      *string        `json:"reason_code,omitempty"`
	Detail          *string        `json:"detail,omitempty"`
	ResponseMessage map[string]any `json:"response_message,omitempty"`
}

type SendResult struct {
	OperationID string            `json:"operation_id"`
	MessageID   string            `json:"message_id"`
	MessageIDs  []string          `json:"message_ids"`
	Outcomes    []DeliveryOutcome `json:"outcomes"`
}

func ParseMessageClass(value string) *MessageClass {
	switch MessageClass(value) {
	case MessageSend, MessageAck, MessageFail, MessageCapabilities, MessageCompensate:
		parsed := MessageClass(value)
		return &parsed
	default:
		return nil
	}
}

type EnvelopeInput struct {
	Sender           string
	Recipients       []string
	MessageClass     MessageClass
	ContextID        string
	ExpiresInSeconds int
	OperationID      string
	Namespace        string
	CorrelationID    string
	InReplyTo        string
	CryptoSuite      string
}

func BuildEnvelope(input EnvelopeInput) (Envelope, error) {
	now := time.Now().UTC()
	expires := now.Add(time.Duration(maxInt(input.ExpiresInSeconds, 1)) * time.Second)
	operationID := input.OperationID
	if operationID == "" {
		operationID = uuid.NewString()
	}
	envelope := Envelope{
		ACPVersion:   ACPVersion,
		MessageClass: input.MessageClass,
		MessageID:    uuid.NewString(),
		OperationID:  operationID,
		Timestamp:    now.Format(time.RFC3339),
		ExpiresAt:    expires.Format(time.RFC3339),
		Sender:       input.Sender,
		Recipients:   append([]string{}, input.Recipients...),
		ContextID:    input.ContextID,
		CryptoSuite:  firstNonBlankString(input.CryptoSuite, DefaultCryptoSuite),
	}
	if input.CorrelationID != "" {
		envelope.CorrelationID = &input.CorrelationID
	}
	if input.InReplyTo != "" {
		envelope.InReplyTo = &input.InReplyTo
	}
	if strings.TrimSpace(input.Namespace) != "" {
		namespace := strings.TrimSpace(input.Namespace)
		envelope.Namespace = &namespace
	}
	if err := ValidateEnvelope(envelope); err != nil {
		return Envelope{}, err
	}
	return envelope, nil
}

func ValidateEnvelope(envelope Envelope) error {
	if envelope.Sender == "" {
		return ValidationError("Envelope sender is required")
	}
	if len(envelope.Recipients) == 0 {
		return ValidationError("Envelope recipients must not be empty")
	}
	parsedTimestamp, err := time.Parse(time.RFC3339, envelope.Timestamp)
	if err != nil {
		return ValidationError("Envelope timestamps must be RFC3339 strings")
	}
	parsedExpires, err := time.Parse(time.RFC3339, envelope.ExpiresAt)
	if err != nil {
		return ValidationError("Envelope timestamps must be RFC3339 strings")
	}
	if !parsedExpires.After(parsedTimestamp) {
		return ValidationError("Envelope expires_at must be after timestamp")
	}
	return nil
}

func IsExpired(envelope Envelope) bool {
	expires, err := time.Parse(time.RFC3339, envelope.ExpiresAt)
	if err != nil {
		return true
	}
	return !expires.After(time.Now().UTC())
}

func ParseAcpMessage(messageMap map[string]any) (AcpMessage, error) {
	serialized, err := json.Marshal(messageMap)
	if err != nil {
		return AcpMessage{}, ValidationError("invalid ACP message payload")
	}
	var message AcpMessage
	if err := json.Unmarshal(serialized, &message); err != nil {
		return AcpMessage{}, ValidationError("invalid ACP message payload")
	}
	if err := ValidateEnvelope(message.Envelope); err != nil {
		return AcpMessage{}, err
	}
	return message, nil
}

func MessageToMap(message AcpMessage) (map[string]any, error) {
	serialized, err := json.Marshal(message)
	if err != nil {
		return nil, ValidationError("unable to serialize ACP message")
	}
	var asMap map[string]any
	if err := json.Unmarshal(serialized, &asMap); err != nil {
		return nil, ValidationError("unable to serialize ACP message")
	}
	return asMap, nil
}

func BuildAckPayload(receivedMessageID, status string) map[string]any {
	return map[string]any{
		"status":              status,
		"received_message_id": receivedMessageID,
	}
}

func BuildFailPayload(reasonCode, detail string, retriable bool) map[string]any {
	return map[string]any{
		"reason_code": reasonCode,
		"detail":      detail,
		"retriable":   retriable,
	}
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func firstNonBlankString(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}
